### ISSUE TITLE
nixos/noXlibs breaks gtk2 / gtk3 / gtk4

### ISSUE THREAD

**Issue by dminuoso** [NixOS/nixpkgs#135810]
:EDIT:

I changed the title from `noXlibs breaks gtk ...` to `noXlibs breaks jre` since the real problem is that once noXlibs is enabled, it seems one cannot have any form of jre anymore.

### Describe the bug
When building a nixos profile with
```
{
  config.environment.noXlibs = true;
  config.environment.systemPackages = [ pkgs.gtk2 ];
}
```

You get a failing builder with:
```
configure: error: Package requirements (cairo-xlib >= 1.6) were not met:

No package 'cairo-xlib' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables CAIRO_BACKEND_CFLAGS
and CAIRO_BACKEND_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
builder for '/nix/store/97p47bwvg0x8wgf7a9ri0blcc5yq9p04-gtk+-2.24.32.drv' failed with exit code 1
error: build of '/nix/store/97p47bwvg0x8wgf7a9ri0blcc5yq9p04-gtk+-2.24.32.drv' on 'ssh://builder' failed: builder for '/nix/store/97p47bwvg0x8wgf7a9ri0blcc5yq9p04-gtk+-2.24.32.drv' failed with exit code 1
builder for '/nix/store/97p47bwvg0x8wgf7a9ri0blcc5yq9p04-gtk+-2.24.32.drv' failed with exit code 1
cannot build derivation '/nix/store/1j2vbnrcxdjz9m5jdr53fssv6jsfcil5-system-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/inzihrq5b385g1yj7fw24c99bd8jzxfl-nixos-system-unifi-21.05pre-git.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/3lysliyhi5ffaiwq6qfllmhrajr0ygbm-morph.drv': 1 dependencies couldn't be built
error: build of '/nix/store/3lysliyhi5ffaiwq6qfllmhrajr0ygbm-morph.drv' failed
Error while running `/run/current-system/sw/bin/nix-build /run/user/1000/morph-977101554/morph/eval-machines.nix -A machines --arg networkExpr /home/dminuoso/private/network/network.nix --argstr argsFile /run/user/1000/morph-726395305/morph-args.json --out-link /run/user/1000/morph-726395305/result --option max-jobs 0 ...`: See above.%     
```

Similar errors for gtk3 and gtk4.

This is somewhat annoying since I generally like `<nixpkgs/nixos/modules/profiles/minimal.nix>` for my servers, but gtk is a dependency of openjdk that I need for some things.

### Steps To Reproduce

See above.

### Notify maintainers
@lopsided98 (introduced the cairo modification in no-x-libs)
@lovek323 @7c6f434c (gtk maintainers)

### Metadata

N/A

This happens on 21.05 or on nixos-unstable.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - gtk2
  - gtk3
  - gtk4
# a list of nixos modules affected by the problem
module:
```

**Comment by taku0** [comment]
`adoptopenjdk-bin` uses `autoPatchelfHook` to link X11 libraries.  We may able to use `patchelf --remove-needed` to tell `autoPatchelfHook` not requireing X11, or just ignore errors by `autoPatchelfIgnoreMissingDeps`.  `adoptopenjdk-bin` works just fine without X11 libraries unless classes using X11 is used.

**Comment by dminuoso** [comment]
We are barking up the wrong tree here. After a good nights sleep I realized that it's just gtk3 that's failing. So all we need to do is remove the GTK dependency from the bootstrap JDK, and use a headless jre, and all seems well.

```nix
  jre8_trueHeadless = (pkgs.openjdk8.override
    { openjdk8-bootstrap = pkgs.openjdk8-bootstrap.override { gtkSupport = false; };
      headless = true;
      enableGnome2 = false;
    }).jre;
```

Seems to do the job. @edwtjo is it sensible to disable gtkSupport on the bootstrap JDK? I don't fully understand what the bootstrap JDK needs to be able to do. Is this just needed for building OpenJDK?

**Comment by aakropotkin** [comment]
This is currently preventing me from updating my NixOS boxes as well. 4 different machines actually which share similar configs. 

`gtk4` basically hangs. `nix` dumps logs griping about `meson` and `ninja` ( perhaps a battle tested build system might be more suitable for a project like `gtk` 😡 ).

I doubt this is a Nix issue. My guess based on the logs is that `ninja` attempts `rm -r $PWD` or a similar directory deletion which has open file descriptors.

**Comment by mohe2015** [comment]
> This is currently preventing me from updating my NixOS boxes as well. 4 different machines actually which share similar configs. 
> 
> `gtk4` basically hangs. `nix` dumps logs griping about `meson` and `ninja` ( perhaps a battle tested build system might be more suitable for a project like `gtk` 😡 ).
> 
> I doubt this is a Nix issue. My guess based on the logs is that `ninja` attempts `rm -r $PWD` or a similar directory deletion which has open file descriptors. 

Of you really think this is the same issue couldn't you just disable noXLibs? And why does this only occur since a recent update? I think the issue mentioned here exists since forever.

**Comment by aakropotkin** [comment]
> > This is currently preventing me from updating my NixOS boxes as well. 4 different machines actually which share similar configs.
> > `gtk4` basically hangs. `nix` dumps logs griping about `meson` and `ninja` ( perhaps a battle tested build system might be more suitable for a project like `gtk` rage ).
> > I doubt this is a Nix issue. My guess based on the logs is that `ninja` attempts `rm -r $PWD` or a similar directory deletion which has open file descriptors.
> 
> Of you really think this is the same issue couldn't you just disable noXLibs? And why does this only occur since a recent update? I think the issue mentioned here exists since forever.

You're right my issue is different. The debug output has a lot of similarities so this popped up in my search results; likely just because it's the most recent `NixOS` + `gtkX` issue.

**Comment by asbachb** [comment]
> I don't fully understand what the bootstrap JDK needs to be able to do. Is this just needed for building OpenJDK?

Yeah the bootstrap jdk (nowerdays called boot jdk) is needed to build the jdk itself. 

see https://github.com/openjdk/jdk17u/blob/master/doc/building.md#boot-jdk-requirements

**Comment by Thomasdezeeuw** [comment]
This is still a problem stale bot

**Comment by SuperSandro2000** [comment]
> I also don't like this but I think this is more or less "intended" behaviour and probably not easily fixable.

Yes and I personally stumbled over this option a couple of times and rather would remove it to reduce unexpected build failures in containers.

**Comment by rnhmjoj** [comment]
It also breaks `nixos-rebuild build-vm`, for some reason. I don't see why an option of the guest system would affect the build of the VM on the host, but it does. To reproduce it's enough to do:
```
imports = [ (modulesPath + "/profiles/minimal.nix") ];
```
and `nixos-rebuild build-vm`.

**Comment by SuperSandro2000** [comment]
There are mixed different problems into this issue, hence I am closing it to untangle them.

That GTK is not working without x11 or on a headless system, is acceptable in my opinion as it is a pure GUI library and already disabled in multiple places in noXlibs.

That the openjdk bootstrapping breaks with it, is unrelated to the original issue and I would suggest to just booststrap without GUI which would also speed the build process up.

Also the build-vm comment above this one is probably a totally separate issue not belonging into this one.

**Comment by rnhmjoj** [comment]
> Also the build-vm comment above this one is probably a totally separate issue not belonging into this one.

Maybe, but the build fails with exactly the same error reported here.

**Comment by dminuoso** [comment]
@SuperSandro2000 

These two issues are precisely related. It is because the bootstrap JDK induces an (unneeded) gnome dependency that config.environment.noXlibs conflicts with the presence of a JRE.

I've made a proposal that went unnoticed by the maintainer.

> We are barking up the wrong tree here. After a good nights sleep I realized that it's just gtk3 that's failing. So all we need to do is remove the GTK dependency from the bootstrap JDK, and use a headless jre, and all seems well.
> 
> ```nix
>   jre8_trueHeadless = (pkgs.openjdk8.override
>     { openjdk8-bootstrap = pkgs.openjdk8-bootstrap.override { gtkSupport = false; };
>       headless = true;
>       enableGnome2 = false;
>     }).jre;
> ```
> 
> Seems to do the job. @edwtjo is it sensible to disable gtkSupport on the bootstrap JDK? I don't fully understand what the bootstrap JDK needs to be able to do. Is this just needed for building OpenJDK?

**Comment by SuperSandro2000** [comment]
I think I fixed that with #213812
